sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala

```diff
@@ -670,6 +670,15 @@ object SQLConf {
     .booleanConf
     .createWithDefault(false)
 
+  val ADAPTIVE_EXECUTION_APPLY_FINAL_STAGE_SHUFFLE_OPTIMIZATIONS =
+    buildConf("spark.sql.adaptive.applyFinalStageShuffleOptimizations")
```

Comment:
@ulysses-you is this useful even for non-table-cache queries?

Reply:
bucket table write ? but I think people would specify partition number explicitly if there is a shuffle on bucket column. I can't find other case.

Reply:
yea this can be a use case. People may want to fully control the partitioning of the final write stage, which can affect number of files.